src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,923 @@
+                 m_voxelGridDimensions[2];
+  for (int i = 0; i < gridSize; i++) {
+    if (func(boolGrids_[boolGridName].get()[i])) {
+      voxelSet.push_back(reverseHash(i));
+    }
+  }
+}
+
+void VoxelGrid::fillVoxelSetFromIntGrid(std::vector<Mn::Vector3i>& voxelSet,
+                                        const std::string& intGridName,
+                                        bool (*func)(int)) {
+  assert(intGrids_.find(intGridName) != intGrids_.end());
+  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                 m_voxelGridDimensions[2];
+  for (int i = 0; i < gridSize; i++) {
+    if (func(intGrids_[intGridName].get()[i])) {
+      voxelSet.push_back(reverseHash(i));
+    }
+  }
+}
+
+void VoxelGrid::fillVoxelSetFromFloatGrid(std::vector<Mn::Vector3i>& voxelSet,
+                                          const std::string& floatGridName,
+                                          bool (*func)(float)) {
+  assert(floatGrids_.find(floatGridName) != floatGrids_.end());
+  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                 m_voxelGridDimensions[2];
+  for (int i = 0; i < gridSize; i++) {
+    if ((*func)(floatGrids_[floatGridName].get()[i])) {
+      voxelSet.push_back(reverseHash(i));
+    }
+  }
+}
+
+void VoxelGrid::fillVoxelSetFromVector3Grid(std::vector<Mn::Vector3i>& voxelSet,
+                                            const std::string& vector3GridName,
+                                            bool (*func)(Mn::Vector3)) {
+  assert(vector3Grids_.find(vector3GridName) != vector3Grids_.end());
+  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                 m_voxelGridDimensions[2];
+  for (int i = 0; i < gridSize; i++) {
+    if (func(vector3Grids_[vector3GridName].get()[i])) {
+      voxelSet.push_back(reverseHash(i));
+    }
+  }
+}
+
+// 6D SDF - labels each cell as interior (-inf), exterior (+inf), or boundary
+// (0)
+void VoxelGrid::generateInteriorExteriorVoxelGrid() {
+  // create 6 bool grids
+  addBoolGrid("negXShadow");
+  addBoolGrid("posXShadow");
+  addBoolGrid("negYShadow");
+  addBoolGrid("posYShadow");
+  addBoolGrid("negZShadow");
+  addBoolGrid("posZShadow");
+  // fill each grids with ray cast
+  bool hit = false;
+  int ind = 0;
+  std::string gridName = "InteriorExterior";
+  // X axis ray casts
+  for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+    for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+      // fill negX
+      hit = false;
+      ind = m_voxelGridDimensions[0] - 1;
+      while (ind--) {
+        if (hit) {
+          setBoolVoxelByIndex(Mn::Vector3i(ind, j, k), true, "negXShadow");
+        } else if (getBoolVoxelByIndex(Mn::Vector3i(ind, j, k))) {
+          hit = true;
+          setBoolVoxelByIndex(Mn::Vector3i(ind, j, k), true, "negXShadow");
+        }
+      }
+      // fill posX
+      hit = false;
+      ind = -1;
+      while (++ind < m_voxelGridDimensions[0]) {
+        if (hit) {
```

Comment:
This could be simplified with something like ``` hit || (hit = getBoolVoxelByIndex(Mn::Vector3i(ind, j, k)) ``` or maybe 
```c++
hit |= getBoolVoxelByIndex(Mn::Vector3i(ind, j, k));
if (hit)  setBoolVoxelByIndex(Mn::Vector3i(ind, j, k), true, "posXShadow");
```